spark3/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead3.java

```diff
@@ -0,0 +1,502 @@
+          .queryName("testtablewithdelete")
+          .outputMode(OutputMode.Append())
+          .start();
+
+      try {
+        streamingQuery.processAllAvailable();
+        Assert.assertTrue(false); // should be unreachable
+      } catch (Exception exception) {
+        Assert.assertTrue(exception instanceof StreamingQueryException);
+        Assert.assertTrue(((StreamingQueryException) exception).cause() instanceof IllegalStateException);
+      }
+
+    } finally {
+      for (StreamingQuery query : spark.streams().active()) {
+        query.stop();
+      }
+    }
+  }
+
+  private static List<SimpleRecord> processMicroBatch(DataStreamWriter<Row> singleBatchWriter, String viewName)
+      throws TimeoutException, StreamingQueryException {
+    StreamingQuery streamingQuery = singleBatchWriter.start();
+    streamingQuery.awaitTermination();
+
+    return spark.sql(String.format("select * from %s", viewName))
+        .as(Encoders.bean(SimpleRecord.class))
+        .collectAsList();
+  }
+
+  /**
+   * get test data - a list of records per snapshot
+   */
+  private static List<List<SimpleRecord>> getTestDataForMultipleSnapshots() {
+    return Lists.newArrayList(
+        Lists.newArrayList(
+            new SimpleRecord(1, "one"),
+            new SimpleRecord(2, "two"),
+            new SimpleRecord(3, "three")),
+        Lists.newArrayList(
+            new SimpleRecord(4, "four"),
+            new SimpleRecord(5, "five")),
+        Lists.newArrayList(
+            new SimpleRecord(6, "six"),
+            new SimpleRecord(7, "seven"))
+    );
+  }
+
+  /**
+   * appends each list as a Snapshot on the iceberg table at the given location.
+   * accepts a list of lists - each list representing data per snapshot.
+   */
+  private static void appendData(List<List<SimpleRecord>> data, File location) {
+    // generate multiple snapshots
+    for (List<SimpleRecord> l : data) {
+      appendData(l, location, "parquet");
+    }
+  }
+
+  private static void appendData(List<SimpleRecord> data, File location, String fileFormat) {
+    Dataset<Row> df = spark.createDataFrame(data, SimpleRecord.class);
+    df.select("id", "data").write()
+        .format("iceberg")
+        .option("write-format", fileFormat)
+        .mode("append")
+        .save(location.toString());
+  }
+
+  private static List<SimpleRecord> processStreamTillEnd(Dataset<Row> df) throws TimeoutException {
+    StreamingQuery streamingQuery = df.writeStream()
+        .format("memory")
+        .queryName("test12")
+        .outputMode(OutputMode.Append())
+        .start();
+    streamingQuery.processAllAvailable();
+    return spark.sql("select * from test12")
+        .as(Encoders.bean(SimpleRecord.class))
+        .collectAsList();
+  }
+
+  /**
+   * gets test data - to be used for multiple write batches
+   * each batch inturn will have multiple snapshots
+   */
+  private static List<List<List<SimpleRecord>>> getTestDataForMultipleWritesWithMultipleSnapshots() {
```

Comment:
Why isn't this a static final variable?

Reply:
changed to `static final`. 
i like to keep them as methods in general - so that if we need to generate data - we can do so. here, i don't feel the need.